MetricsTest.availableViaJmx

http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/4881/testReport/junit/com.hazelcast.jet.core.metrics/MetricsTest/availableViaJmx/


```
name: total not published
```
```
java.lang.AssertionError: name: total not published
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at com.hazelcast.jet.core.metrics.JmxMetricsChecker.getMetricValue(JmxMetricsChecker.java:55)
	at com.hazelcast.jet.core.metrics.MetricsTest.availableViaJmx(MetricsTest.java:422)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

Also failed in http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/4926/testReport/junit/com.hazelcast.jet.core.metrics/MetricsTest/availableViaJmx/ with the same stack trace. Adding the output in this message. 

Output

```
Started Running Test: usingServiceAsync
08:49:38,283  INFO |usingServiceAsync| - [system] Thread-173 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] 
	o  o   O   o---o o--o o      o-o   O    o-o  o-O-o     o--o  o       O  o-O-o o--o  o-o  o--o  o   o 
	|  |  / \     /  |    |     /     / \  |       |       |   | |      / \   |   |    o   o |   | |\ /| 
	O--O o---o  -O-  O-o  |    O     o---o  o-o    |       O--o  |     o---o  |   O-o  |   | O-Oo  | O | 
	|  | |   |  /    |    |     \    |   |     |   |       |     |     |   |  |   |    o   o |  \  |   | 
	o  o o   o o---o o--o O---o  o-o o   o o--o    o       o     O---o o   o  o   o     o-o  o   o o   o
08:49:38,283  INFO |usingServiceAsync| - [system] Thread-173 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
08:49:38,283  INFO |usingServiceAsync| - [system] Thread-173 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Hazelcast Platform 5.0-SNAPSHOT (20210506 - f17156f) starting at [127.0.0.1]:5701
08:49:38,283  INFO |usingServiceAsync| - [system] Thread-173 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Cluster name: dev
08:49:38,283  INFO |usingServiceAsync| - [system] Thread-173 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Jet extension is enabled.
08:49:38,292  WARN |usingServiceAsync| - [CPSubsystem] Thread-173 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.
08:49:38,297  INFO |usingServiceAsync| - [JetService] Thread-173 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Setting number of cooperative threads and default parallelism to 8
08:49:38,298  INFO |usingServiceAsync| - [Diagnostics] Thread-173 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
08:49:38,298  INFO |usingServiceAsync| - [LifecycleService] Thread-173 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5701 is STARTING
08:49:38,299  INFO |usingServiceAsync| - [ClusterService] Thread-173 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] 

Members {size:1, ver:1} [
	Member [127.0.0.1]:5701 - 8e06e499-f732-4c26-8efd-e09537ef85ed this
]

08:49:38,299 DEBUG |usingServiceAsync| - [JobCoordinationService] hz.compassionate_kirch.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
08:49:38,299  INFO |usingServiceAsync| - [ManagementService] Thread-173 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Hazelcast JMX agent enabled.
08:49:38,299  INFO |usingServiceAsync| - [LifecycleService] Thread-173 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5701 is STARTED
08:49:38,399 DEBUG |usingServiceAsync| - [JobCoordinationService] hz.compassionate_kirch.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
08:49:38,422  INFO |usingServiceAsync| - [PartitionStateManager] hz.compassionate_kirch.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Initializing cluster partition table arrangement...
08:49:38,455 DEBUG |usingServiceAsync| - [Planner] hz.compassionate_kirch.cached.thread-1 - Watermarks in the pipeline will be throttled to 1000
08:49:38,503 DEBUG |usingServiceAsync| - [JobRepository] hz.compassionate_kirch.cached.thread-2 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Job cleanup took 3ms
08:49:38,555  INFO |usingServiceAsync| - [JobCoordinationService] hz.compassionate_kirch.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Starting job 0624-766c-9000-0001 based on submit request
08:49:38,920  INFO |usingServiceAsync| - [MasterJobContext] hz.compassionate_kirch.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Didn't find any snapshot to restore for job '0624-766c-9000-0001', execution 0624-766c-9001-0001
08:49:38,920  INFO |usingServiceAsync| - [MasterJobContext] hz.compassionate_kirch.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Start executing job '0624-766c-9000-0001', execution 0624-766c-9001-0001, execution graph in DOT format:
digraph DAG {
	"items" [localParallelism=1];
	"add-timestamps" [localParallelism=1];
	"mapUsingServiceAsync" [localParallelism=8];
	"noop" [localParallelism=1];
	"items" -> "add-timestamps" [label="isolated", queueSize=1024];
	"add-timestamps" -> "mapUsingServiceAsync" [queueSize=1024];
	"mapUsingServiceAsync" -> "noop" [queueSize=1024];
}
HINT: You can use graphviz or http://viz-js.com to visualize the printed graph.
08:49:38,920 DEBUG |usingServiceAsync| - [MasterJobContext] hz.compassionate_kirch.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Building execution plan for job '0624-766c-9000-0001', execution 0624-766c-9001-0001
08:49:38,921 DEBUG |usingServiceAsync| - [MasterJobContext] hz.compassionate_kirch.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Built execution plans for job '0624-766c-9000-0001', execution 0624-766c-9001-0001
08:49:38,970  INFO |availableViaJmx| - [JetTestSupport] Thread-139 - Terminating instanceFactory in JetTestSupport.@After
08:49:38,971  INFO |availableViaJmx| - [LifecycleService] Thread-175 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
08:49:38,972  WARN |availableViaJmx| - [Node] Thread-175 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Terminating forcefully...
08:49:38,972  INFO |availableViaJmx| - [Node] Thread-175 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Shutting down connection manager...
08:49:38,973  INFO |availableViaJmx| - [Node] Thread-175 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Shutting down node engine...
08:49:38,974 DEBUG |availableViaJmx| - [MasterJobContext] ForkJoinPool.commonPool-worker-6 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Execution of job '0624-766a-d780-0001', execution 0624-766a-d781-0001 has failures: [[127.0.0.1]:5701=com.hazelcast.core.HazelcastInstanceNotActiveException: Hazelcast instance is not active!]
08:49:38,974 DEBUG |availableViaJmx| - [MasterJobContext] hz.sweet_kirch.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Sending CompleteExecutionOperation for job '0624-766a-d780-0001', execution 0624-766a-d781-0001
08:49:38,974 ERROR |availableViaJmx| - [MasterJobContext] ForkJoinPool.commonPool-worker-6 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] job '0624-766a-d780-0001', execution 0624-766a-d781-0001: some TerminateExecutionOperation invocations failed, execution might remain stuck: [MemberInfo{address=[127.0.0.1]:5701, uuid=ac6638a2-b048-4c3e-9b34-4b19ad1e3281, liteMember=false, memberListJoinVersion=1}=com.hazelcast.core.HazelcastInstanceNotActiveException: Hazelcast instance is not active!]
08:49:38,974 ERROR |availableViaJmx| - [MasterJobContext] ForkJoinPool.commonPool-worker-2 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] job '0624-766a-d780-0001', execution 0624-766a-d781-0001: some CompleteExecutionOperation invocations failed, execution resources might leak: [MemberInfo{address=[127.0.0.1]:5701, uuid=ac6638a2-b048-4c3e-9b34-4b19ad1e3281, liteMember=false, memberListJoinVersion=1}=com.hazelcast.core.HazelcastInstanceNotActiveException: Hazelcast instance is not active!]
08:49:38,974 DEBUG |availableViaJmx| - [JobCoordinationService] hz.sweet_kirch.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Scheduling restart on master for job 0624-766a-d780-0001
08:49:38,974 DEBUG |availableViaJmx| - [JobExecutionService] hz.sweet_kirch.jet.cooperative.thread-6 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Execution of job '0624-766a-d780-0001', execution 0624-766a-d781-0001 completed with failure
java.util.concurrent.CompletionException: java.util.concurrent.CancellationException
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[?:1.8.0_251]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) ~[?:1.8.0_251]
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:607) ~[?:1.8.0_251]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:591) ~[?:1.8.0_251]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_251]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990) ~[?:1.8.0_251]
	at com.hazelcast.jet.impl.util.NonCompletableFuture.internalCompleteExceptionally(NonCompletableFuture.java:59) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$ExecutionTracker.taskletDone(TaskletExecutionService.java:469) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.lambda$run$0(TaskletExecutionService.java:356) ~[classes/:?]
	at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891) [?:1.8.0_251]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:356) [classes/:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_251]
Caused by: java.util.concurrent.CancellationException
	at java.util.concurrent.CompletableFuture.cancel(CompletableFuture.java:2276) ~[?:1.8.0_251]
	at com.hazelcast.jet.impl.execution.ExecutionContext.terminateExecution(ExecutionContext.java:233) ~[classes/:?]
	at com.hazelcast.jet.impl.JobExecutionService.cancelAndComplete(JobExecutionService.java:184) ~[classes/:?]
	at com.hazelcast.jet.impl.JobExecutionService.lambda$cancelAllExecutions$4(JobExecutionService.java:158) ~[classes/:?]
	at java.util.concurrent.ConcurrentHashMap$ValuesView.forEach(ConcurrentHashMap.java:4707) ~[?:1.8.0_251]
	at com.hazelcast.jet.impl.JobExecutionService.cancelAllExecutions(JobExecutionService.java:154) ~[classes/:?]
	at com.hazelcast.jet.impl.JobExecutionService.shutdown(JobExecutionService.java:143) ~[classes/:?]
	at com.hazelcast.jet.impl.JetService.shutdown(JetService.java:182) ~[classes/:?]
	at com.hazelcast.spi.impl.servicemanager.impl.ServiceManagerImpl.shutdownService(ServiceManagerImpl.java:307) ~[classes/:?]
	at com.hazelcast.spi.impl.servicemanager.impl.ServiceManagerImpl.shutdown(ServiceManagerImpl.java:298) ~[classes/:?]
	at com.hazelcast.spi.impl.NodeEngineImpl.shutdown(NodeEngineImpl.java:515) ~[classes/:?]
	at com.hazelcast.instance.impl.Node.shutdownServices(Node.java:591) ~[classes/:?]
	at com.hazelcast.instance.impl.Node.shutdown(Node.java:530) ~[classes/:?]
	at com.hazelcast.instance.impl.LifecycleServiceImpl.shutdown(LifecycleServiceImpl.java:101) ~[classes/:?]
	at com.hazelcast.instance.impl.LifecycleServiceImpl.terminate(LifecycleServiceImpl.java:89) ~[classes/:?]
	at com.hazelcast.test.mocknetwork.TestNodeRegistry.shutdown(TestNodeRegistry.java:123) ~[test-classes/:?]
	at com.hazelcast.test.mocknetwork.TestNodeRegistry.terminate(TestNodeRegistry.java:113) ~[test-classes/:?]
	at com.hazelcast.test.TestHazelcastInstanceFactory.terminateAll(TestHazelcastInstanceFactory.java:331) ~[test-classes/:?]
	at com.hazelcast.client.test.TestHazelcastFactory.terminateAll(TestHazelcastFactory.java:183) ~[test-classes/:?]
	at com.hazelcast.jet.JetTestInstanceFactory.terminateAll(JetTestInstanceFactory.java:129) ~[test-classes/:?]
	at com.hazelcast.jet.core.JetTestSupport.lambda$shutdownFactory$0(JetTestSupport.java:82) ~[test-classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_251]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_251]
	... 1 more
08:49:38,975 DEBUG |availableViaJmx| - [JobExecutionService] hz.sweet_kirch.jet.cooperative.thread-6 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Completing job '0624-766a-d780-0001', execution 0624-766a-d781-0001 locally. Reason: Node is shutting down
08:49:38,975 DEBUG |availableViaJmx| - [JobExecutionService] hz.sweet_kirch.jet.cooperative.thread-6 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Completed execution of job '0624-766a-d780-0001', execution 0624-766a-d781-0001
08:49:38,979  INFO |availableViaJmx| - [NodeExtension] Thread-175 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Destroying node NodeExtension.
08:49:38,979  INFO |availableViaJmx| - [Node] Thread-175 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Hazelcast Shutdown is completed in 7 ms.
08:49:38,979  INFO |availableViaJmx| - [LifecycleService] Thread-175 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN
BuildInfo right after availableViaJmx(com.hazelcast.jet.core.metrics.MetricsTest): BuildInfo{version='5.0-SNAPSHOT', build='20210506', buildNumber=20210506, revision=f17156f, enterprise=false, serializationVersion=1, jet=JetBuildInfo{version='5.0-SNAPSHOT', build='20210506', revision='f17156f'}}
Hiccups measured while running test 'availableViaJmx(com.hazelcast.jet.core.metrics.MetricsTest):'
08:49:35, accumulated pauses: 907 ms, max pause: 333 ms, pauses over 1000 ms: 0
```

--- a/hazelcast/src/test/java/com/hazelcast/jet/core/metrics/JmxMetricsChecker.java
+++ b/hazelcast/src/test/java/com/hazelcast/jet/core/metrics/JmxMetricsChecker.java
@@ -52,7 +52,7 @@ class JmxMetricsChecker {
         Set<ObjectName> publishedDescriptors = platformMBeanServer
                 .queryMBeans(new ObjectName(PREFIX + ":*"), null)
                 .stream().map(ObjectInstance::getObjectName).collect(toSet());
-        assertTrue("name: " + metricName + " not published", publishedDescriptors.contains(descriptor));
+        assertTrue("metric '" + metricName + "' not published", publishedDescriptors.contains(descriptor));
         return (long) platformMBeanServer.getAttribute(descriptor, metricName);
     }
 

--- a/hazelcast/src/test/java/com/hazelcast/jet/core/metrics/MetricsTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/jet/core/metrics/MetricsTest.java
@@ -71,13 +71,14 @@ public class MetricsTest extends JetTestSupport {
 
     private final Pipeline pipeline = Pipeline.create();
     private JetInstance instance;
+    private Config config;
 
     @Before
     public void before() {
-        Config config = new Config();
+        config = smallInstanceConfig();
         config.setProperty("hazelcast.jmx", "true");
         config.getMetricsConfig().setCollectionFrequencySeconds(1);
-        instance = createJetMembers(config, 1)[0];
+        instance = createJetMember(config);
 
         TestProcessors.reset(1);
     }
@@ -415,7 +416,7 @@ public class MetricsTest extends JetTestSupport {
 
         String instanceName = instance.getHazelcastInstance().getName();
         long sum = 0;
-        int availableProcessors = Runtime.getRuntime().availableProcessors();
+        int availableProcessors = config.getJetConfig().getInstanceConfig().getCooperativeThreadCount();
         for (int i = 0; i < availableProcessors; i++) {
             JmxMetricsChecker jmxMetricsChecker = new JmxMetricsChecker(instanceName, job,
                     "vertex=fused(filter, map)", "procType=TransformP", "proc=" + i, "user=true");

